hips/hip-9999.md

```diff
@@ -0,0 +1,254 @@
+---
+hip: 9999
+title: "New annotations for displaying hook output"
+authors: [ "Ian Zink <ian@replicated.com>" ]
+created: "2023-01-26"
+type: "feature"
+status: "draft"
+---
+
+## Abstract
+
+This proposes a new annotation to indicate that the output from the hook should be displayed to the user.
+
+## Motivation
+
+The one motivation for this HIP is the ability to display preflight checks before the main chart installs. This provides vital feedback to the user as to why a Helm chart can not be successfully installed.
+
+Often it is important to verify that the Kubernetes cluster you are deploying a Helm chart into has certain properties. You might need to know that it has an Ingress controller, a certain amount of ephemeral storage, memory, or CPUs available. You might want to validate the service key they provided was correct or that that database they entered is reachable. Letting a Helm install and/or upgrade fail, and then make the user debug this to see why it failed is a poor user experience. These things can all be done with checks enabled by the hook proposed in this HIP.
+
+Another common motivation for this HIP is to indicate that database migration output running as a hook should be visible to the user in the case of failure.
+
+In general, allowing chart developers to run jobs and present that feedback directly to the users could also open up additional use cases beyond just the preflight use case that motivated this HIP.  I could imagine scenarios where maybe CVE warnings are presented or specific upgrade feedback is presented instead of just a Helm install failure.
+
+## Rationale
+
+There are other ways that this could be implemented. For example, we could have a separate preflight hook type. However, this new hook type wouldn't be handled at all by previous versions of Helm. With this design, it requires minimal changes to Helm and allows for backwards compatibility.
+
+Another strategy could be for Helm to include Troubleshoot.sh as a dependent library, but this could result in too tight of a coupling between the projects and lower overall flexibility and adaptability. 
+
+## Specification
+
+Templates may include the following annotations on Jobs or Pods:
+    
+```yaml
+        "helm.sh/hook": pre-install, pre-upgrade
+        "helm.sh/hook-output-log-policy": hook-failed # or hook-succeeded or hook-failed,hook-succeeded
+```
+
+
+`helm.sh/hook-output-log-policy` would indicate that Helm should display the output of the Job to the user.
+
+Additionally, a new user flag should be created `--no-log-output`  that would skip the output of logs.
+
+Additionally, there will be a new item added to the action SDK configuration to allow SDK consumers to get the output. 
+By default this output will be written to stdout, but an SDK consumer can overwrite the HookOutputFunc to provide a custom writer.
```

Comment:
I don't think the SDK should write to StdOut. StdErr is where logs usually go. Consider a service, like Flux, running this update. How can things be low impact on them? What about writing to `io.Discard` by default. They set a writer if they want to capture the output. This is how they can opt-in to getting that output.